mhs/mhs/common/workflow/sync_async.py

```diff
@@ -33,18 +33,24 @@ def __init__(self, config_manager: configuration_manager.ConfigurationManager,
         self.transmission = transmission
         self.party_id = party_id
 
-    def prepare_message(self, interaction_name: str, content: str) -> typing.Tuple[bool, str, str]:
+    def prepare_message(self, interaction_name: str, content: str, message_id: typing.Optional[str] = None) \
+            -> typing.Tuple[bool, str, str]:
         """Prepare a message to be sent for the specified interaction. Wraps the provided content if required.
 
         :param interaction_name: The name of the interaction the message is related to.
         :param content: The message content to be sent.
+        :param message_id: message id to use in the message header. If not supplied, we generate a new message id.
+        The purpose of this param is to be used for duplicate elimination when sending async messages.
         :return: A tuple containing:
         1. A flag indicating whether this message should be sent asynchronously
         2. the ID of the message, if an asynchronous response is expected, otherwise None
         3. The message to send
         """
         interaction_details = self._get_interaction_details(interaction_name)
 
+        if message_id is not None:
+            interaction_details[ebxml_envelope.MESSAGE_ID] = message_id
```

Comment:
This will apply if `message_id` is an empty string, would there ever be a situation where we want an empty message id or should we just not have the pair?

Reply:
I doubt we'd want an empty message id, but I don't think we should do any validation here as this isn't the interface we expose to the outside world, and it feels like it might be overkill here. However, I think we should validate _somewhere_. I'm not sure if this needs to be part of this task, last week @gareth-allan mentioned to me that we've got some other task for sorting out the HTTP interface, so I'd imagine we'd do HTTP request validation at that time?

Reply:
Fair point, we'll have to remember to do this somewhere else that's appropriate

Reply:
I've mentioned this to Colin to put into the relevant story (as more of a general thing, not just for message id).